### DISCUSSION TITLE
Improve offline support 🛌

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [meteor/meteor#11656] [Feature Requests]
This recently came up in a discussion on Slack: https://meteor-community.slack.com/archives/CN350MY1G/p1632871131110600?thread_ts=1632857949.109300&cid=CN350MY1G
And was lightly touched in: https://github.com/meteor/meteor/discussions/11598

With Minimongo there is a touch of offline support. This could be taken further and improved. First step most likely being the ability to not loose offline only changes when the app is closed and then improved syncing once connection is restored. The final step possibly being that you could create offline-first Meteor app.

Some steps have been taken in this direction with [ground:db](https://github.com/GroundMeteor/db).

There is also [Hoodie](https://github.com/hoodiehq) which uses PouchDB and CouchDB as a pair to achieve this. [Read more](https://www.twilio.com/blog/2018/03/practical-introduction-pwa-node-hoodie-offline-first.html)

Update on some additional solutions that I saw:
* [Replicache](https://replicache.dev/)
* [RxDB](https://rxdb.info/)
* [PartyKit](https://www.partykit.io/)
* [Query](https://github.com/TanStack/query) <= maybe integration with this?
* [SWR](https://swr.vercel.app/)
* [Watermelon DB](https://watermelondb.dev/docs) - Rocket.chat uses this, but it is React specific
* [ClientDB](https://clientdb.dev/)
* [Tinybase](https://tinybase.org/)
* [PouchDB](https://docs.yjs.dev/)

For merging data there might data, there is:
* [Yjs](https://docs.yjs.dev/)
* [Automerge](https://automerge.org/)

What are your thoughts/ideas?

**Comment by renanccastro** [comment]
I've created a mechanism back in the days where I stored methods calls serialized in the IndexedDB and replayed them when the user had a connection again.
The good thing about this approach is that we can always apply optimistic updates to get a responsive UI right away, even offline, and then sync the changes.

The challenge is in the server, to reconcile the changes if there was some previous state change in the meanwhile, but this is something that varies in terms of business decisions.

Maybe we could create a package now using this, what do you think?

**Reply by StorytellerCZ** [reply]
What about `jam:offline` with `jam:method`?

**Reply by xet7** [reply]
Or https://github.com/jamauro/offline ?

**Comment by jamauro** [comment]
Can we revive this and include as a candidate for the roadmap? I would love to see this brought into Meteor as a core feature. There’s a move towards offline / local-first data sync for the speed and reliability benefits. See Linear and Muse as some examples. I think apps will continue moving in this direction and Meteor could position itself as the go to platform. 

Maybe @banjerluke’s gist is a good starting point: https://forums.meteor.com/t/is-grounddb-the-best-for-offline-storage/56654/16?u=jam

cc @fredmaiaarantes

**Comment by paulincai** [comment]
Hi @jamauro,

I personally think this should not be in the core simply because it cannot be in the core. I'll explain why.
There are multiple front-ends now with Meteor. There is probably very little difference to what the server looks like when moving from one front-end to another.
While offline is mostly a front-end thing (and sync requires some server code) strategies for offline are different and developers want to treat this differently.
The core of Meteor holds the most general "truth". Other versions of the truth can and should (IMHO) be built on top of the general truth. Offline-first is not a general truth because it is too subjected to the developers preference.

I, for instance, use React with Redux. Redux has something called redux-persist (https://blog.logrocket.com/persist-state-redux-persist-redux-toolkit-react/) which contains absolutely everything one may need to build a fully offline platform. You can have a look at the type of stores they integrate with: https://github.com/rt2zz/redux-persist#storage-engines

I definitely would not want Meteor to come and tell me to do offline in a certain way.

I support more Meteor documentation, blogging, opinion exchange rather than adding components to the core.

Also, in the Meteor community there are multiple "teams". I am personally in Team React and Team Server-Only-Bundle. They way offline works in my team and my demand for offline is not the same as for Team Svelte or Team Blaze. Even within Team React, the way I want to resolute a data sync between offline and server is not the same in all projects or from a developer/company to another.

I really think offline is like your CSS library. You cannot put Bootstrap or MUI in the core.

I hope you understand that I am not against you or your opinion, I just have my own and given the right arguments, I could even push some code to an official or community offline component for Meteor. I just don't see the need for it.

**Reply by StorytellerCZ** [reply]
One thing that is the same across all the front-ends is minimongo, so I think some sort of a sync strategy in situation when the app goes offline for any reason. In that case methods and changes to minimongo can still be fired, so adding something to catch those and then sync with server I think is the appropriate first step in this. MongoDB guys are also looking into this for web now as well, they already have it for mobile via [Realm](https://realm.io/) if I remember correctly from MongoDB World 2022.

**Reply by paulincai** [reply]
Across all front-ends you MAY have minimongo for whoever uses it and for those you never know how much data you have in your local DB which leads me back to my earlier statement, it really depends on the content of the app and developer how you build offline.

After I red your comment, I picked  up my phone and set it in flight mode and started to open native apps. Most of the ones I tried (Youtube, Gmail, Strava, Garmin, Waze, Messenger) offer some offline experience and in 2-3 clicks you start getting error messages. The same thing I get with my PWAs, nothing more.

It is true, you can send some messages while offline and they are synced when online but would I do this via minimongo? Perhaps never, because you cannot trust minimongo. You can trust a local DB but not minimimongo because by nature it is reactive.

Optimistic UI, if I am not wrong, also depends on having methods defined both server and client side. Some projects like ... all my projects do not expose methods on the client side for security reasons (and don't use publications).

I think for offline the most efficient way would be to fully rely on a local DB, perhaps IndexedDB or some other local store to write changes like a log for e very limited number of actions such as change state of a radio button, send a message, do a sign out. Media is cached by its cacheing policies anyway.

The thing is that I am so biased into Redux as global store that I find it hard now to think that people use trackers or methods inside components or views. For those cases offline could be a nightmare.

The more I write the more I realize that the actual problem is having a reliable non-volatile global store (such as Redux) that can get json or objects from minimongo on an offline event and serve the app via functions. With this data store you don't call methods, instead, you log the actions such as:
If Meteor connected:
 * call method A with this params
 * call method B with this params
When offline all these changes at UI level are operated in the local store via functions. When connected, those methods in the log (part of the client sync API) are called 1 by 1 or bulk.

We, here, do not believe in minimongo and also we don't believe that, at least for us, minimongo is the solution for offline, now or in the future. It could be for some but even in those cases, the API would be too complex to be worth it. The more features and users a web app has, the less likely to have an offline success via minimongo.
1-2 screens and all data published with the user, minimongo will work 100% for offline. But ... 

Perhaps I should correct my earlier statement. Offline would be nice to have in one way or another for whoever wants to use it as is (as it would be built for Meteor), but perhaps should not be in the core because the way to resolve sync is too dependent on the app content and while any content can be viewed offline from a store (minimongo or something else), only certain actions can be offline, perhaps only "writes" and not all writes.
Only writes of your own data that is exclusively owned by you such as your profile, your messages, your notifications, that cannot be altered by others like bookings, purchases, confirmations of attendance with limited seats, create items in a list with unique names etc.

**Reply by paulincai** [reply]
From Realm (these are the things I referred to as - only certain actions can be offline).  Most of the Sync is app content  and collection schema dependent and if you factor in aggregations or Grapher things become even uglier :

https://www.mongodb.com/docs/atlas/app-services/sync/learn/overview/#std-label-conflict-resolution-overview


"Partition key values and data access rules determine which users can read or write which data in your database.

When you use [Flexible Sync](https://www.mongodb.com/docs/atlas/app-services/sync/data-access-patterns/flexible-sync/#std-label-flexible-sync), you must:

Specify queryable fields
Configure data access rules and default roles
Client applications can query the queryable fields you configure to determine which objects to sync. Then, App Services applies rules and default roles to determine whether users can read or write the objects that match the query."

**Reply by banjerluke** [reply]
For what it's worth, I added full offline support to my Meteor web/Cordova app a few years ago – I wrote some about my implementation in the forum thread @jamauro linked to – and at least for me, I can't really think of anything that could be added to the core that would have been of use to me. Different kinds of data necessitate different syncing strategies; I use at least three different ones in my app. FYI I'm using minimongo and am on Team Blaze / Team Svelte.

**Comment by paulincai** [comment]
An example of queueing middleware. When offline, dispatched actions (methods) optimistically. Supports both sync and async. When online unloads the queue. https://github.com/mathieudutour/redux-queue-offline
An another version of offline from the Redux world: https://github.com/redux-offline/redux-offline

**Comment by xet7** [comment]
# New packages

## 1)  jam:offline - An easy way to give your Meteor app offline capabilities

https://forums.meteor.com/t/new-package-jam-offline-an-easy-way-to-give-your-meteor-app-offline-capabilities/61874

## 2) jamauro:offline - An easy way to give your Meteor app offline capabilities and make it feel instant

https://github.com/jamauro/offline

https://forums.meteor.com/t/meteor-roadmap-update-august-27th-2024/62103/6

**Reply by StorytellerCZ** [reply]
Those two are the same package.
